extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,238 @@
+}
+
+function getTokensForPosition(tokens: Token[], position: vscode.Position): Token[] {
+	let enclosingTokens = tokens.filter(token => token.map && (token.map[0] <= position.line && token.map[1] > position.line) && isBlockElement(token));
+
+	if (enclosingTokens.length === 0) {
+		return [];
+	}
+
+	let sortedTokens = enclosingTokens.sort((token1, token2) => (token2.map[1] - token2.map[0]) - (token1.map[1] - token1.map[0]));
+	return sortedTokens;
+}
+
+function getHeadersForPosition(toc: TocEntry[], position: vscode.Position): { headers: TocEntry[], headerOnThisLine: boolean } {
+	let enclosingHeaders = toc.filter(header => header.location.range.start.line <= position.line && header.location.range.end.line >= position.line);
+	let sortedHeaders = enclosingHeaders.sort((header1, header2) => (header1.line - position.line) - (header2.line - position.line));
+	let onThisLine = toc.find(header => header.line === position.line) !== undefined;
+	return {
+		headers: sortedHeaders,
+		headerOnThisLine: onThisLine
+	};
+}
+
+function isBlockElement(token: Token): boolean {
+	return !['list_item_close', 'paragraph_close', 'bullet_list_close', 'inline', 'heading_close', 'heading_open'].includes(token.type);
+}
+
+function createHeaderRange(isClosestHeaderToPosition: boolean, onHeaderLine: boolean, header?: TocEntry, parent?: vscode.SelectionRange, childStart?: vscode.Position): vscode.SelectionRange | undefined {
+	if (header) {
+		let contentRange = new vscode.Range(header.location.range.start.translate(1), header.location.range.end);
+		let headerPlusContentRange = header.location.range;
+		let partialContentRange = childStart && isClosestHeaderToPosition ? contentRange.with(undefined, childStart) : undefined;
+		if (onHeaderLine && isClosestHeaderToPosition && childStart) {
+			return new vscode.SelectionRange(header.location.range.with(undefined, childStart), new vscode.SelectionRange(header.location.range, parent));
+		} else if (onHeaderLine && isClosestHeaderToPosition) {
+			return new vscode.SelectionRange(header.location.range, parent);
+		} else if (parent && parent.range.contains(headerPlusContentRange)) {
+			if (partialContentRange) {
+				return new vscode.SelectionRange(partialContentRange, new vscode.SelectionRange(contentRange, (new vscode.SelectionRange(headerPlusContentRange, parent))));
+			} else {
+				return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(headerPlusContentRange, parent));
+			}
+		} else if (partialContentRange) {
+			return new vscode.SelectionRange(partialContentRange, new vscode.SelectionRange(contentRange, (new vscode.SelectionRange(headerPlusContentRange))));
+		} else {
+			return new vscode.SelectionRange(contentRange, new vscode.SelectionRange(headerPlusContentRange));
+		}
+	} else {
+		return undefined;
+	}
+}
+
+function createBlockRange(document: vscode.TextDocument, cursorLine: number, block?: Token, parent?: vscode.SelectionRange): vscode.SelectionRange | undefined {
```

Comment:
Can you make `block` be non nullable here? 

Up where `blockTokens.shift()` is called you may need to add a not null assertion (`blockTokens.shift()!`) but that's ok since you check that the array is not empty before calling `shift`